src/ucp/proto/proto_init.c

```diff
@@ -196,196 +115,178 @@ ucp_proto_perf_envelope_make(const ucs_linear_func_t *funcs, uint64_t funcs_num,
+            /* Count only send completion time without waiting for a response */
+            ((op_attr_mask & UCP_OP_ATTR_FLAG_FAST_CMPL) &&
+             !(params->flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE)));
+}
 
-        ucs_trace("range[%d] %s" UCP_PROTO_PERF_FUNC_TYPES_FMT,
-                  caps->num_ranges,
-                  ucs_memunits_range_str(range_start, range->max_length,
-                                         range_str, sizeof(range_str)),
-                  UCP_PROTO_PERF_FUNC_TYPES_ARG(range->perf));
+static ucs_status_t
+ucp_proto_init_add_tl_perf(const ucp_proto_common_init_params_t *params,
+                           const ucp_proto_common_tl_perf_t *tl_perf,
+                           ucp_proto_perf_node_t *const tl_perf_node,
+                           size_t range_start, size_t range_end,
+                           ucp_proto_perf_t *perf)
+{
+    ucp_proto_perf_factors_t perf_factors = UCP_PROTO_PERF_FACTORS_INITIALIZER;
+    const double send_overhead            = tl_perf->send_pre_overhead +
+                                            tl_perf->send_post_overhead;
+    uint32_t op_attr_mask;
 
-        stage_node = ucp_proto_perf_node_new_compose("stage", "");
+    ucs_trace("caps" UCP_PROTO_TIME_FMT(send_pre_overhead)
+              UCP_PROTO_TIME_FMT(send_post_overhead)
+              UCP_PROTO_TIME_FMT(recv_overhead) UCP_PROTO_TIME_FMT(latency)
+              UCP_PROTO_TIME_FMT(sys_latency),
+              UCP_PROTO_TIME_ARG(tl_perf->send_pre_overhead),
+              UCP_PROTO_TIME_ARG(tl_perf->send_post_overhead),
+              UCP_PROTO_TIME_ARG(tl_perf->recv_overhead),
+              UCP_PROTO_TIME_ARG(tl_perf->latency),
+              UCP_PROTO_TIME_ARG(tl_perf->sys_latency));
 
-        ucs_carray_for_each(stage_elem, stages, num_stages) {
-            if (ucp_proto_perf_range_is_zero(*stage_elem)) {
-                continue;
-            }
+    op_attr_mask = ucp_proto_select_op_attr_unpack(
+            params->super.select_param->op_attr);
 
-            /* range->node ---> stage_node ---> [ stage{i}->node ... ] */
-            ucp_proto_perf_node_add_child(stage_node, (*stage_elem)->node);
-        }
+    perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_CPU].c += send_overhead;
+    perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY].c   += tl_perf->latency;
+    perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY].c   += tl_perf->sys_latency;
 
-        ucp_proto_perf_node_own_child(range->node, &stage_node);
+    /* Add remote CPU TL overhead */
+    if (!ucp_proto_init_skip_recv_overhead(params, op_attr_mask)) {
+        perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_CPU].c +=
+                tl_perf->recv_overhead;
+    }
 
-        ++caps->num_ranges;
-        range_start = range->max_length + 1;
+    /* With non-zcopy fast-completion we don't count transport time */
+    if (!(op_attr_mask & UCP_OP_ATTR_FLAG_FAST_CMPL) ||
+        (params->flags & UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY)) {
```

Comment:
should we also do it for `params->flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE`?

Reply:
actually, for UCP_PROTO_COMMON_INIT_FLAG_RESPONSE we should count the BW as REMOTE_TL instead of LOCAL_TL, but we can do such tuning in next PR IMO.

Reply:
ok